File: rtls-kmp/src/androidMain/kotlin/com/rtls/kmp/SqliteLocationStore.kt
package com.rtls.kmp

import android.content.Context
import android.database.Cursor
import android.database.sqlite.SQLiteDatabase
import android.database.sqlite.SQLiteOpenHelper
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

/**
 * SQLite-backed LocationStore using SQLiteOpenHelper so the platform handles
 * DB lifecycle and threading. Only considers points with sent_at IS NULL AND failed_at IS NULL
 * as pending (failed points are not re-uploaded indefinitely).
 */
class SqliteLocationStore(private val context: Context) : LocationStore, SentPointsPrunableLocationStore, BidirectionalLocationStore {

    private val helper by lazy { RtlsDbHelper(context) }

    private val pendingWhere = "sent_at IS NULL AND failed_at IS NULL"

    override suspend fun insert(points: List<LocationPoint>) = withContext(Dispatchers.IO) {
        if (points.isEmpty()) return@withContext
        val db = helper.writableDatabase
        db.beginTransaction()
        try {
            points.forEach { p ->
                db.execSQL(
                    """INSERT OR IGNORE INTO location_points
                        (id, user_id, device_id, recorded_at, lat, lng, horizontal_accuracy, vertical_accuracy, altitude, speed, course, sent_at, failed_at, error_message)
                        VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, NULL, NULL, NULL)""",
                    arrayOf(
                        p.id, p.userId, p.deviceId, p.recordedAtMs, p.lat, p.lng,
                        p.horizontalAccuracy, p.verticalAccuracy, p.altitude, p.speed, p.course
                    )
                )
            }
            db.setTransactionSuccessful()
        } finally {
            db.endTransaction()
        }
    }

    override suspend fun fetchPendingPoints(limit: Int): List<LocationPoint> = withContext(Dispatchers.IO) {
        val db = helper.readableDatabase
        val c = db.rawQuery(
            "SELECT id, user_id, device_id, recorded_at, lat, lng, horizontal_accuracy, vertical_accuracy, altitude, speed, course FROM location_points WHERE $pendingWhere ORDER BY recorded_at ASC LIMIT ?",
            arrayOf(limit.toString())
        )
        val list = mutableListOf<LocationPoint>()
        while (c.moveToNext()) {
            list.add(
                LocationPoint(
                    id = c.getString(0),
                    userId = c.getString(1),
                    deviceId = c.getString(2),
                    recordedAtMs = c.getLong(3),
                    lat = c.getDouble(4),
                    lng = c.getDouble(5),
                    horizontalAccuracy = if (c.isNull(6)) null else c.getDouble(6),
                    verticalAccuracy = if (c.isNull(7)) null else c.getDouble(7),
                    altitude = if (c.isNull(8)) null else c.getDouble(8),
                    speed = if (c.isNull(9)) null else c.getDouble(9),
                    course = if (c.isNull(10)) null else c.getDouble(10)
                )
            )
        }
        c.close()
        list
    }

    override suspend fun pendingCount(): Int = withContext(Dispatchers.IO) {
        val db = helper.readableDatabase
        val c = db.rawQuery("SELECT COUNT(*) FROM location_points WHERE $pendingWhere", null)
        c.moveToFirst()
        val n = c.getInt(0)
        c.close()
        n
    }

    override suspend fun oldestPendingRecordedAt(): Long? = withContext(Dispatchers.IO) {
        val db = helper.readableDatabase
        val c = db.rawQuery("SELECT MIN(recorded_at) FROM location_points WHERE $pendingWhere", null)
        val result = if (c.moveToFirst() && !c.isNull(0)) c.getLong(0) else null
        c.close()
        result
    }

    override suspend fun pendingStats(): PendingStats = withContext(Dispatchers.IO) {
        val db = helper.readableDatabase
        val c = db.rawQuery(
            "SELECT COUNT(*), MIN(recorded_at) FROM location_points WHERE $pendingWhere", null
        )
        val stats = if (c.moveToFirst()) {
            PendingStats(c.getInt(0), if (c.isNull(1)) null else c.getLong(1))
        } else {
            PendingStats(0, null)
        }
        c.close()
        stats
    }

    override suspend fun markSent(pointIds: List<String>, sentAtMs: Long) = withContext(Dispatchers.IO) {
        if (pointIds.isEmpty()) return@withContext
        val db = helper.writableDatabase
        for (chunk in pointIds.chunked(999)) {
            val placeholders = chunk.joinToString(",") { "?" }
            val args = arrayOf(sentAtMs.toString()) + chunk.toTypedArray()
            db.execSQL("UPDATE location_points SET sent_at = ? WHERE id IN ($placeholders)", args)
        }
    }

    override suspend fun markFailed(pointIds: List<String>, errorMessage: String) = withContext(Dispatchers.IO) {
        if (pointIds.isEmpty()) return@withContext
        val now = System.currentTimeMillis()
        val db = helper.writableDatabase
        for (chunk in pointIds.chunked(999)) {
            val placeholders = chunk.joinToString(",") { "?" }
            val args = arrayOf(now.toString(), errorMessage) + chunk.toTypedArray()
            db.execSQL("UPDATE location_points SET failed_at = ?, error_message = ? WHERE id IN ($placeholders)", args)
        }
    }

    override suspend fun pruneSentPoints(olderThanRecordedMs: Long) = withContext(Dispatchers.IO) {
        val db = helper.writableDatabase
        db.execSQL(
            "DELETE FROM location_points WHERE sent_at IS NOT NULL AND recorded_at < ?",
            arrayOf(olderThanRecordedMs)
        )
    }

    override suspend fun applyServerChanges(
        items: List<LocationPoint>,
        mergeStrategy: LocationMergeStrategy?,
        serverTimeMs: Long?,
        lastSyncAtMs: Long?
    ) = withContext(Dispatchers.IO) {
        if (items.isEmpty()) return@withContext
        val context = MergeContext(lastSyncAtMs = lastSyncAtMs, serverTimeMs = serverTimeMs)
        val db = helper.writableDatabase
        db.beginTransaction()
        try {
            for (serverItem in items) {
                val local = fetchLocalPoint(db, serverItem.id)
                if (local != null) {
                    if (local == serverItem) continue
                    when (val result = mergeStrategy?.resolve(local, serverItem, context) ?: LocationMergeResult.KeepServer) {
                        is LocationMergeResult.KeepLocal -> { /* no write */ }
                        is LocationMergeResult.KeepServer -> upsertReplica(db, serverItem)
                        is LocationMergeResult.Use -> upsertReplica(db, result.point)
                    }
                } else {
                    upsertReplica(db, serverItem)
                }
            }
            db.setTransactionSuccessful()
        } finally {
            db.endTransaction()
        }
    }

    override suspend fun getLastPullCursor(): SyncCursor? = withContext(Dispatchers.IO) {
        val db = helper.readableDatabase
        val c = db.rawQuery("SELECT value FROM sync_metadata WHERE key = 'pull_cursor' LIMIT 1", null)
        val result = if (c.moveToFirst() && !c.isNull(0)) SyncCursor(c.getBlob(0)) else null
        c.close()
        result
    }

    override suspend fun setLastPullCursor(cursor: SyncCursor?) = withContext(Dispatchers.IO) {
        val db = helper.writableDatabase
        if (cursor != null) {
            db.execSQL(
                "INSERT OR REPLACE INTO sync_metadata(key, value) VALUES ('pull_cursor', ?)",
                arrayOf(cursor.value)
            )
        } else {
            db.delete("sync_metadata", "key = ?", arrayOf("pull_cursor"))
        }
        Unit
    }

    private fun fetchLocalPoint(db: SQLiteDatabase, id: String): LocationPoint? {
        fetchReplicaPoint(db, id)?.let { return it }
        return fetchPendingPointById(db, id)
    }

    private fun fetchReplicaPoint(db: SQLiteDatabase, id: String): LocationPoint? {
        val c = db.rawQuery(
            "SELECT id, user_id, device_id, recorded_at, lat, lng, horizontal_accuracy, vertical_accuracy, altitude, speed, course FROM sync_replica_location_points WHERE id = ?",
            arrayOf(id)
        )
        val point = if (c.moveToFirst()) rowToPoint(c) else null
        c.close()
        return point
    }

    private fun fetchPendingPointById(db: SQLiteDatabase, id: String): LocationPoint? {
        val c = db.rawQuery(
            "SELECT id, user_id, device_id, recorded_at, lat, lng, horizontal_accuracy, vertical_accuracy, altitude, speed, course FROM location_points WHERE id = ? AND $pendingWhere",
            arrayOf(id)
        )
        val point = if (c.moveToFirst()) rowToPoint(c) else null
        c.close()
        return point
    }

    private fun rowToPoint(c: Cursor): LocationPoint =
        LocationPoint(
            id = c.getString(0),
            userId = c.getString(1),
            deviceId = c.getString(2),
            recordedAtMs = c.getLong(3),
            lat = c.getDouble(4),
            lng = c.getDouble(5),
            horizontalAccuracy = if (c.isNull(6)) null else c.getDouble(6),
            verticalAccuracy = if (c.isNull(7)) null else c.getDouble(7),
            altitude = if (c.isNull(8)) null else c.getDouble(8),
            speed = if (c.isNull(9)) null else c.getDouble(9),
            course = if (c.isNull(10)) null else c.getDouble(10)
        )

    private fun upsertReplica(db: SQLiteDatabase, point: LocationPoint) {
        val now = System.currentTimeMillis()
        db.execSQL(
            """INSERT OR REPLACE INTO sync_replica_location_points
                (id, user_id, device_id, recorded_at, lat, lng, horizontal_accuracy, vertical_accuracy, altitude, speed, course, updated_at)
                VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)""",
            arrayOf(
                point.id, point.userId, point.deviceId, point.recordedAtMs, point.lat, point.lng,
                point.horizontalAccuracy, point.verticalAccuracy, point.altitude, point.speed, point.course,
                now
            )
        )
    }

    private class RtlsDbHelper(context: Context) : SQLiteOpenHelper(
        context,
        "rtlsync.db",
        null,
        2
    ) {
        init {
            setWriteAheadLoggingEnabled(true)
        }

        override fun onConfigure(db: SQLiteDatabase) {
            super.onConfigure(db)
            db.execSQL("PRAGMA synchronous = NORMAL")
        }

        override fun onCreate(db: SQLiteDatabase) {
            db.execSQL("""
                CREATE TABLE location_points (
                    id TEXT PRIMARY KEY,
                    user_id TEXT NOT NULL,
                    device_id TEXT NOT NULL,
                    recorded_at INTEGER NOT NULL,
                    lat REAL NOT NULL,
                    lng REAL NOT NULL,
                    horizontal_accuracy REAL,
                    vertical_accuracy REAL,
                    altitude REAL,
                    speed REAL,
                    course REAL,
                    sent_at INTEGER,
                    failed_at INTEGER,
                    error_message TEXT
                )
            """.trimIndent())
            db.execSQL(
                "CREATE INDEX idx_location_points_pending ON location_points(sent_at, failed_at, recorded_at)"
            )
        }

        override fun onUpgrade(db: SQLiteDatabase, oldVersion: Int, newVersion: Int) {
            if (oldVersion < 2) {
                db.execSQL("""
                    CREATE TABLE IF NOT EXISTS sync_replica_location_points (
                        id TEXT PRIMARY KEY,
                        user_id TEXT NOT NULL,
                        device_id TEXT NOT NULL,
                        recorded_at INTEGER NOT NULL,
                        lat REAL NOT NULL,
                        lng REAL NOT NULL,
                        horizontal_accuracy REAL,
                        vertical_accuracy REAL,
                        altitude REAL,
                        speed REAL,
                        course REAL,
                        updated_at INTEGER NOT NULL
                    )
                """.trimIndent())
                db.execSQL("""
                    CREATE TABLE IF NOT EXISTS sync_metadata (
                        key TEXT PRIMARY KEY,
                        value BLOB
                    )
                """.trimIndent())
            }
        }
    }
}

File: Sources/RTLSData/SQLiteLocationStore.swift
import Foundation
import RTLSCore
import SQLite3
import os.log

public actor SQLiteLocationStore: LocationStore, SentPointsPrunableLocationStore, BidirectionalLocationStore {
    private nonisolated let SQLITE_TRANSIENT = unsafeBitCast(-1, to: sqlite3_destructor_type.self)

    private enum SyncStatus: Int {
        case pending = 0
        case sent = 1
        case failed = 2
    }

    private let db: SQLiteDatabase

    public init(databaseURL: URL) async throws {
        self.db = try SQLiteDatabase(url: databaseURL)
        try migrateIfNeeded()
    }

    private func migrateIfNeeded() throws {
        try db.exec("""
        CREATE TABLE IF NOT EXISTS schema_version (
            version INTEGER NOT NULL
        );
        """)

        let current = try currentSchemaVersion()
        if current == 0 {
            try db.exec("""
            CREATE TABLE IF NOT EXISTS location_points (
                id TEXT PRIMARY KEY,
                user_id TEXT NOT NULL,
                device_id TEXT NOT NULL,
                recorded_at_ms INTEGER NOT NULL,
                lat REAL NOT NULL,
                lng REAL NOT NULL,
                hacc REAL,
                vacc REAL,
                altitude REAL,
                speed REAL,
                course REAL,
                sync_status INTEGER NOT NULL,
                last_error TEXT,
                sent_at_ms INTEGER
            );
            """)
            try db.exec("""
            CREATE INDEX IF NOT EXISTS idx_location_points_status_time
            ON location_points(sync_status, recorded_at_ms);
            """)

            try setSchemaVersion(1)
        }
        if current < 2 {
            try db.exec("""
            CREATE TABLE IF NOT EXISTS sync_replica_location_points (
                id TEXT PRIMARY KEY,
                user_id TEXT NOT NULL,
                device_id TEXT NOT NULL,
                recorded_at_ms INTEGER NOT NULL,
                lat REAL NOT NULL,
                lng REAL NOT NULL,
                hacc REAL,
                vacc REAL,
                altitude REAL,
                speed REAL,
                course REAL,
                updated_at_ms INTEGER NOT NULL
            );
            """)
            try db.exec("""
            CREATE TABLE IF NOT EXISTS sync_metadata (
                key TEXT PRIMARY KEY,
                value BLOB
            );
            """)
            try setSchemaVersion(2)
        }
    }

    private func currentSchemaVersion() throws -> Int {
        let stmt = try db.prepare("SELECT version FROM schema_version LIMIT 1;")
        defer { sqlite3_finalize(stmt) }

        if sqlite3_step(stmt) == SQLITE_ROW {
            return Int(sqlite3_column_int(stmt, 0))
        }
        return 0
    }

    private func setSchemaVersion(_ version: Int) throws {
        try db.exec("DELETE FROM schema_version;")
        let stmt = try db.prepare("INSERT INTO schema_version(version) VALUES (?);")
        defer { sqlite3_finalize(stmt) }

        guard sqlite3_bind_int(stmt, 1, Int32(version)) == SQLITE_OK else {
            throw SQLiteError.bind(db.lastErrorMessage())
        }
        guard sqlite3_step(stmt) == SQLITE_DONE else {
            throw SQLiteError.step(db.lastErrorMessage())
        }
    }

    public func insert(points: [LocationPoint]) async throws {
        guard !points.isEmpty else { return }

        try db.exec("BEGIN TRANSACTION;")
        do {
            let sql = """
            INSERT OR IGNORE INTO location_points(
                id, user_id, device_id, recorded_at_ms, lat, lng, hacc, vacc, altitude, speed, course,
                sync_status, last_error, sent_at_ms
            ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, NULL, NULL);
            """

            let stmt = try db.prepare(sql)
            defer { sqlite3_finalize(stmt) }

            for p in points {
                sqlite3_reset(stmt)
                sqlite3_clear_bindings(stmt)

                try bindText(stmt, index: 1, value: p.id.uuidString)
                try bindText(stmt, index: 2, value: p.userId)
                try bindText(stmt, index: 3, value: p.deviceId)
                try bindInt64(stmt, index: 4, value: Self.dateToMs(p.recordedAt))
                try bindDouble(stmt, index: 5, value: p.coordinate.latitude)
                try bindDouble(stmt, index: 6, value: p.coordinate.longitude)
                try bindNullableDouble(stmt, index: 7, value: p.horizontalAccuracy)
                try bindNullableDouble(stmt, index: 8, value: p.verticalAccuracy)
                try bindNullableDouble(stmt, index: 9, value: p.altitude)
                try bindNullableDouble(stmt, index: 10, value: p.speed)
                try bindNullableDouble(stmt, index: 11, value: p.course)
                try bindInt(stmt, index: 12, value: SyncStatus.pending.rawValue)

                guard sqlite3_step(stmt) == SQLITE_DONE else {
                    throw SQLiteError.step(db.lastErrorMessage())
                }
            }

            try db.exec("COMMIT;")
        } catch {
            // Attempt rollback and log any errors from both the original operation and rollback
            do {
                try db.exec("ROLLBACK;")
            } catch let rollbackError {
                // Log rollback failure - this indicates serious DB corruption
                os_log(.error, "RTLS: Transaction failed and rollback also failed: %{public}@. Original error: %{public}@",
                       rollbackError.localizedDescription, error.localizedDescription)
                // Throw composite error to preserve both errors
                throw SQLiteError.transactionFailed(original: error, rollback: rollbackError)
            }
            throw error
        }
    }

    public func fetchPendingPoints(limit: Int) async throws -> [LocationPoint] {
        guard limit > 0 else { return [] }

        let sql = """
        SELECT
            id, user_id, device_id, recorded_at_ms, lat, lng, hacc, vacc, altitude, speed, course
        FROM location_points
        WHERE sync_status = ?
        ORDER BY recorded_at_ms ASC
        LIMIT ?;
        """

        let stmt = try db.prepare(sql)
        defer { sqlite3_finalize(stmt) }

        try bindInt(stmt, index: 1, value: SyncStatus.pending.rawValue)
        try bindInt(stmt, index: 2, value: limit)

        var points: [LocationPoint] = []
        while sqlite3_step(stmt) == SQLITE_ROW {
            points.append(decodePointRow(stmt))
        }
        return points
    }

    public func fetchPendingPoints(userId: String, deviceId: String, limit: Int) async throws -> [LocationPoint] {
        guard limit > 0 else { return [] }

        let sql = """
        SELECT
            id, user_id, device_id, recorded_at_ms, lat, lng, hacc, vacc, altitude, speed, course
        FROM location_points
        WHERE sync_status = ?
          AND user_id = ?
          AND device_id = ?
        ORDER BY recorded_at_ms ASC
        LIMIT ?;
        """

        let stmt = try db.prepare(sql)
        defer { sqlite3_finalize(stmt) }

        try bindInt(stmt, index: 1, value: SyncStatus.pending.rawValue)
        try bindText(stmt, index: 2, value: userId)
        try bindText(stmt, index: 3, value: deviceId)
        try bindInt(stmt, index: 4, value: limit)

        var points: [LocationPoint] = []
        while sqlite3_step(stmt) == SQLITE_ROW {
            points.append(decodePointRow(stmt))
        }
        return points
    }

    public func fetchRecentPoints(userId: String, deviceId: String, limit: Int) async throws -> [LocationPoint] {
        guard limit > 0 else { return [] }

        let sql = """
        SELECT
            id, user_id, device_id, recorded_at_ms, lat, lng, hacc, vacc, altitude, speed, course
        FROM location_points
        WHERE user_id = ?
          AND device_id = ?
        ORDER BY recorded_at_ms DESC
        LIMIT ?;
        """

        let stmt = try db.prepare(sql)
        defer { sqlite3_finalize(stmt) }

        try bindText(stmt, index: 1, value: userId)
        try bindText(stmt, index: 2, value: deviceId)
        try bindInt(stmt, index: 3, value: limit)

        var points: [LocationPoint] = []
        while sqlite3_step(stmt) == SQLITE_ROW {
            points.append(decodePointRow(stmt))
        }
        return points
    }

    public func pendingCount() async throws -> Int {
        let sql = "SELECT COUNT(1) FROM location_points WHERE sync_status = ?;"
        let stmt = try db.prepare(sql)
        defer { sqlite3_finalize(stmt) }

        try bindInt(stmt, index: 1, value: SyncStatus.pending.rawValue)
        guard sqlite3_step(stmt) == SQLITE_ROW else { return 0 }
        return Int(sqlite3_column_int(stmt, 0))
    }

    public func oldestPendingRecordedAt() async throws -> Date? {
        let sql = """
        SELECT recorded_at_ms
        FROM location_points
        WHERE sync_status = ?
        ORDER BY recorded_at_ms ASC
        LIMIT 1;
        """
        let stmt = try db.prepare(sql)
        defer { sqlite3_finalize(stmt) }

        try bindInt(stmt, index: 1, value: SyncStatus.pending.rawValue)
        guard sqlite3_step(stmt) == SQLITE_ROW else { return nil }
        return Self.msToDate(columnInt64(stmt, index: 0))
    }

    public func pendingStats() async throws -> PendingStats {
        let sql = "SELECT COUNT(1), MIN(recorded_at_ms) FROM location_points WHERE sync_status = ?;"
        let stmt = try db.prepare(sql)
        defer { sqlite3_finalize(stmt) }

        try bindInt(stmt, index: 1, value: SyncStatus.pending.rawValue)
        guard sqlite3_step(stmt) == SQLITE_ROW else { return PendingStats(count: 0, oldestRecordedAt: nil) }
        let count = Int(sqlite3_column_int(stmt, 0))
        let oldest: Date? = (sqlite3_column_type(stmt, 1) == SQLITE_NULL) ? nil : Self.msToDate(columnInt64(stmt, index: 1))
        return PendingStats(count: count, oldestRecordedAt: oldest)
    }

    public func markSent(pointIds: [UUID], sentAt: Date) async throws {
        try mark(pointIds: pointIds, status: .sent, sentAt: sentAt, errorMessage: nil)
    }

    public func markFailed(pointIds: [UUID], errorMessage: String) async throws {
        try mark(pointIds: pointIds, status: .failed, sentAt: nil, errorMessage: errorMessage)
    }

    // MARK: - BidirectionalLocationStore

    public func applyServerChanges(
        _ items: [LocationPoint],
        mergeStrategy: (any LocationMergeStrategy)?,
        serverTime: Date? = nil,
        lastSyncAt: Date? = nil
    ) async throws {
        guard !items.isEmpty else { return }
        let context = MergeContext(lastSyncAt: lastSyncAt, serverTime: serverTime)
        try db.exec("BEGIN TRANSACTION;")
        do {
            for serverItem in items {
                let local = try fetchLocalPoint(id: serverItem.id)
                if let local {
                    if local == serverItem { continue }
                    if let strategy = mergeStrategy {
                        switch strategy.resolve(local: local, server: serverItem, context: context) {
                        case .keepLocal: continue
                        case .keepServer: try upsertReplica(serverItem)
                        case .use(let resolved): try upsertReplica(resolved)
                        }
                    } else {
                        try upsertReplica(serverItem)
                    }
                } else {
                    try upsertReplica(serverItem)
                }
            }
            try db.exec("COMMIT;")
        } catch {
            try? db.exec("ROLLBACK;")
            throw error
        }
    }

    public func getLastPullCursor() async throws -> SyncCursor? {
        let sql = "SELECT value FROM sync_metadata WHERE key = 'pull_cursor' LIMIT 1;"
        let stmt = try db.prepare(sql)
        defer { sqlite3_finalize(stmt) }
        guard sqlite3_step(stmt) == SQLITE_ROW else { return nil }
        guard let blob = sqlite3_column_blob(stmt, 0) else { return nil }
        let count = sqlite3_column_bytes(stmt, 0)
        return SyncCursor(value: Data(bytes: blob, count: Int(count)))
    }

    public func setLastPullCursor(_ cursor: SyncCursor?) async throws {
        let sql = "INSERT OR REPLACE INTO sync_metadata(key, value) VALUES ('pull_cursor', ?);"
        let stmt = try db.prepare(sql)
        defer { sqlite3_finalize(stmt) }
        if let cursor {
            let bytes = [UInt8](cursor.value)
            guard sqlite3_bind_blob(stmt, 1, bytes, Int32(bytes.count), SQLITE_TRANSIENT) == SQLITE_OK else {
                throw SQLiteError.bind(db.lastErrorMessage())
            }
        } else {
            guard sqlite3_bind_null(stmt, 1) == SQLITE_OK else {
                throw SQLiteError.bind(db.lastErrorMessage())
            }
        }
        guard sqlite3_step(stmt) == SQLITE_DONE else {
            throw SQLiteError.step(db.lastErrorMessage())
        }
    }

    private func fetchLocalPoint(id: UUID) throws -> LocationPoint? {
        if let replica = try fetchReplicaPoint(id: id) { return replica }
        return try fetchPendingPoint(id: id)
    }

    private func fetchReplicaPoint(id: UUID) throws -> LocationPoint? {
        let sql = """
        SELECT id, user_id, device_id, recorded_at_ms, lat, lng, hacc, vacc, altitude, speed, course
        FROM sync_replica_location_points WHERE id = ?;
        """
        let stmt = try db.prepare(sql)
        defer { sqlite3_finalize(stmt) }
        try bindText(stmt, index: 1, value: id.uuidString)
        guard sqlite3_step(stmt) == SQLITE_ROW else { return nil }
        return decodePointRow(stmt)
    }

    private func fetchPendingPoint(id: UUID) throws -> LocationPoint? {
        let sql = """
        SELECT id, user_id, device_id, recorded_at_ms, lat, lng, hacc, vacc, altitude, speed, course
        FROM location_points WHERE id = ? AND sync_status = ?;
        """
        let stmt = try db.prepare(sql)
        defer { sqlite3_finalize(stmt) }
        try bindText(stmt, index: 1, value: id.uuidString)
        try bindInt(stmt, index: 2, value: SyncStatus.pending.rawValue)
        guard sqlite3_step(stmt) == SQLITE_ROW else { return nil }
        return decodePointRow(stmt)
    }

    private func upsertReplica(_ point: LocationPoint) throws {
        let sql = """
        INSERT OR REPLACE INTO sync_replica_location_points(
            id, user_id, device_id, recorded_at_ms, lat, lng, hacc, vacc, altitude, speed, course, updated_at_ms
        ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?);
        """
        let stmt = try db.prepare(sql)
        defer { sqlite3_finalize(stmt) }
        let nowMs = Self.dateToMs(Date())
        try bindText(stmt, index: 1, value: point.id.uuidString)
        try bindText(stmt, index: 2, value: point.userId)
        try bindText(stmt, index: 3, value: point.deviceId)
        try bindInt64(stmt, index: 4, value: Self.dateToMs(point.recordedAt))
        try bindDouble(stmt, index: 5, value: point.coordinate.latitude)
        try bindDouble(stmt, index: 6, value: point.coordinate.longitude)
        try bindNullableDouble(stmt, index: 7, value: point.horizontalAccuracy)
        try bindNullableDouble(stmt, index: 8, value: point.verticalAccuracy)
        try bindNullableDouble(stmt, index: 9, value: point.altitude)
        try bindNullableDouble(stmt, index: 10, value: point.speed)
        try bindNullableDouble(stmt, index: 11, value: point.course)
        try bindInt64(stmt, index: 12, value: nowMs)
        guard sqlite3_step(stmt) == SQLITE_DONE else {
            throw SQLiteError.step(db.lastErrorMessage())
        }
    }

    public func pruneSentPoints(olderThan cutoff: Date) async throws {
        let sql = """
        DELETE FROM location_points
        WHERE sync_status = ?
          AND sent_at_ms IS NOT NULL
          AND sent_at_ms < ?;
        """
        let stmt = try db.prepare(sql)
        defer { sqlite3_finalize(stmt) }

        try bindInt(stmt, index: 1, value: SyncStatus.sent.rawValue)
        try bindInt64(stmt, index: 2, value: Self.dateToMs(cutoff))

        guard sqlite3_step(stmt) == SQLITE_DONE else {
            throw SQLiteError.step(db.lastErrorMessage())
        }
    }

    private func mark(pointIds: [UUID], status: SyncStatus, sentAt: Date?, errorMessage: String?) throws {
        guard !pointIds.isEmpty else { return }

        let sentAtMs = sentAt.map(Self.dateToMs)
        let chunkSize = 999
        for chunk in stride(from: 0, to: pointIds.count, by: chunkSize).map({ Array(pointIds[$0..<min($0 + chunkSize, pointIds.count)]) }) {
            let placeholders = chunk.map { _ in "?" }.joined(separator: ",")
            let sql = """
            UPDATE location_points
            SET sync_status = ?, last_error = ?, sent_at_ms = ?
            WHERE id IN (\(placeholders));
            """
            let stmt = try db.prepare(sql)
            defer { sqlite3_finalize(stmt) }

            try bindInt(stmt, index: 1, value: status.rawValue)
            try bindNullableText(stmt, index: 2, value: errorMessage)
            try bindNullableInt64(stmt, index: 3, value: sentAtMs)
            for (offset, id) in chunk.enumerated() {
                try bindText(stmt, index: Int32(4 + offset), value: id.uuidString)
            }

            guard sqlite3_step(stmt) == SQLITE_DONE else {
                throw SQLiteError.step(db.lastErrorMessage())
            }
        }
    }

    // MARK: - Binding helpers

    private func decodePointRow(_ stmt: OpaquePointer?) -> LocationPoint {
        let id = UUID(uuidString: columnText(stmt, index: 0)) ?? UUID()
        let userId = columnText(stmt, index: 1)
        let deviceId = columnText(stmt, index: 2)
        let recordedAt = Self.msToDate(columnInt64(stmt, index: 3))
        let lat = sqlite3_column_double(stmt, 4)
        let lng = sqlite3_column_double(stmt, 5)

        let hacc = columnNullableDouble(stmt, index: 6)
        let vacc = columnNullableDouble(stmt, index: 7)
        let altitude = columnNullableDouble(stmt, index: 8)
        let speed = columnNullableDouble(stmt, index: 9)
        let course = columnNullableDouble(stmt, index: 10)

        return LocationPoint(
            id: id,
            userId: userId,
            deviceId: deviceId,
            recordedAt: recordedAt,
            coordinate: GeoCoordinate(latitude: lat, longitude: lng),
            horizontalAccuracy: hacc,
            verticalAccuracy: vacc,
            altitude: altitude,
            speed: speed,
            course: course
        )
    }

    private func bindText(_ stmt: OpaquePointer?, index: Int32, value: String) throws {
        let result = sqlite3_bind_text(stmt, index, value, -1, SQLITE_TRANSIENT)
        guard result == SQLITE_OK else { throw SQLiteError.bind(db.lastErrorMessage()) }
    }

    private func bindNullableText(_ stmt: OpaquePointer?, index: Int32, value: String?) throws {
        guard let value else {
            let result = sqlite3_bind_null(stmt, index)
            guard result == SQLITE_OK else { throw SQLiteError.bind(db.lastErrorMessage()) }
            return
        }
        try bindText(stmt, index: index, value: value)
    }

    private func bindInt(_ stmt: OpaquePointer?, index: Int32, value: Int) throws {
        let result = sqlite3_bind_int(stmt, index, Int32(value))
        guard result == SQLITE_OK else { throw SQLiteError.bind(db.lastErrorMessage()) }
    }

    private func bindInt64(_ stmt: OpaquePointer?, index: Int32, value: Int64) throws {
        let result = sqlite3_bind_int64(stmt, index, value)
        guard result == SQLITE_OK else { throw SQLiteError.bind(db.lastErrorMessage()) }
    }

    private func bindNullableInt64(_ stmt: OpaquePointer?, index: Int32, value: Int64?) throws {
        guard let value else {
            let result = sqlite3_bind_null(stmt, index)
            guard result == SQLITE_OK else { throw SQLiteError.bind(db.lastErrorMessage()) }
            return
        }
        try bindInt64(stmt, index: index, value: value)
    }

    private func bindDouble(_ stmt: OpaquePointer?, index: Int32, value: Double) throws {
        let result = sqlite3_bind_double(stmt, index, value)
        guard result == SQLITE_OK else { throw SQLiteError.bind(db.lastErrorMessage()) }
    }

    private func bindNullableDouble(_ stmt: OpaquePointer?, index: Int32, value: Double?) throws {
        guard let value else {
            let result = sqlite3_bind_null(stmt, index)
            guard result == SQLITE_OK else { throw SQLiteError.bind(db.lastErrorMessage()) }
            return
        }
        try bindDouble(stmt, index: index, value: value)
    }

    private func columnText(_ stmt: OpaquePointer?, index: Int32) -> String {
        guard let c = sqlite3_column_text(stmt, index) else { return "" }
        return String(cString: c)
    }

    private func columnInt64(_ stmt: OpaquePointer?, index: Int32) -> Int64 {
        sqlite3_column_int64(stmt, index)
    }

    private func columnNullableDouble(_ stmt: OpaquePointer?, index: Int32) -> Double? {
        if sqlite3_column_type(stmt, index) == SQLITE_NULL { return nil }
        return sqlite3_column_double(stmt, index)
    }

    private static func dateToMs(_ date: Date) -> Int64 {
        Int64(date.timeIntervalSince1970 * 1000.0)
    }

    private static func msToDate(_ ms: Int64) -> Date {
        Date(timeIntervalSince1970: TimeInterval(ms) / 1000.0)
    }
}
